tools/jdk/local_java_repository.bzl

```diff
@@ -0,0 +1,69 @@
+"""Java repository implementation
+
+Creates a local repository using jdk.BUILD file.
+
+When java executable is not present it creates a BUILD file with target "jdk"
+displaying an error message.
+"""
+
+def _local_java_repository_impl(repository_ctx):
+    java_home = repository_ctx.attr.java_home
+    java_home_path = repository_ctx.path(java_home)
+    if not java_home_path.exists:
+        fail('The repository\'s path is "%s" (absolute: "%s") ' +
+             "but this directory does not exist." % (java_home, str(java_home_path)))
+
+    repository_ctx.file(
+        "WORKSPACE",
+        "# DO NOT EDIT: automatically generated WORKSPACE file for local_java_repository\n" +
+        "workspace(name = \"{name}\")\n".format(name = repository_ctx.name),
+    )
+
+    extension = ".exe" if repository_ctx.os.name.lower().find("windows") != -1 else ""
+    if java_home_path.get_child("bin").get_child("java" + extension).exists:
+        repository_ctx.file(
+            "BUILD.bazel",
+            repository_ctx.read(repository_ctx.path(repository_ctx.attr._build_file)),
+            False,
+        )
+
+        # Symlink all files
+        for file in repository_ctx.path(java_home).readdir():
+            repository_ctx.symlink(file, file.basename)
+
+        return
+
+    # Java binary does not exist
+    # TODO(ilist): replace error message after toolchain implementation
+    repository_ctx.file(
+        "BUILD.bazel",
+        '''load("@bazel_tools//tools/jdk:local_java_repository.bzl", "failrule")
+
+failrule(
```

Comment:
How does this work? AFAIU this `.bzl` file is evaluated from the WORKSPACE file, so if you call `failure()` there, it will be a repository rule. Also, it seems that the rule is definited below in line 66, but then how is the symbol `failrule` available here?

Reply:
local_hava_repository.bzl file is evaluated twice. From WORKSPACE file (workspace file suffix), where it loads local_java_repository. And from BUILD.bazel file in the newly create workspace (load statement in line 40). Where it loads failrule, which fails when something depends on it.

This is similar to pattern that is used for android repositories. When a version of Android SDK is missing a genrule (bomb) is put there, which displays an error when executed.

Reply:
Argh, you're right, I missed the invocation in line 40. Mind moving the definition of `failrule` to a separate `.bzl` file so that we separate WORKSPACE and non-WORKSPACE Starlark?

Reply:
Done